### PR TITLE
zd15632 - OSS-Fuzz #55810: Out-of-bounds read in certificate parsing

### PR DESCRIPTION
**Pull request by tim-weller-wolfssl**
# Description

**[Fixes zd#15632](https://wolfssl.zendesk.com/agent/tickets/15632)**

Avoids an Out-of-Bounds access caused by parsing of badly-formed ASN1 data, returning an error instead.

# Testing

Added a test for `ParseCert()` API passing badly formed ASN data based on the "reproducer" provided in the ZenDesk ticket.  Ensuring that the calls to `ParseCert()` error out.

### CODE REVIEW

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -1023,7 +1023,14 @@ static int GetOID(const byte* input, word32* inOutIdx, word32* oid,
 static int GetASN_Integer(const byte* input, word32 idx, int length,
                           int positive)
 {
-    if (input[idx] == 0) {
+    /* Check contents consist of one or more octets. */
+    if (length == 0) {
+        #ifdef WOLFSSL_DEBUG_ASN_TEMPLATE
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/asn.c:1028]: GetASN_BitString needs to check length too.

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -1482,6 +1490,10 @@ int GetASN_Items(const ASNItem* asn, ASNGetData *data, int count, int complete,
         }
         /* Move index to start of length. */
         idx++;
+
+        /* Set extra length if indefinite length encoding. */
+        extra = (input[idx] == 0x80 ? 2 : 0);
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:1495]: Change extra to indefExtra.
Indefinite length encoding was not at all handled in original implementation.
'len' will not be correct.
Should be calculated by seeing where the EOC item starts. Best to add check when EOC item (0x00, 0x00) is seen and put length in to the indef length item.
indefExtra is still needed I far as I can see.

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -2131,20 +2143,31 @@ int GetLength_ex(const byte* input, word32* inOutIdx, int* len, word32 maxIdx,
...
 
-        /* Check the number of bytes required are available. */
-        if ((idx + bytes) > maxIdx) {
-            WOLFSSL_MSG("GetLength - bad long length");
+        /* Check the number of bytes required are available.  At least the
+         * number of bytes specified for a long form length or at least two
+         * bytes for the end-of-content octets for an indefinite form length.
+         */
+        if (((idx + bytes) > maxIdx) ||
+            ((bytes == 0) && (idx > (maxIdx - 2)))) {
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:2151]: Don't check bytes == 0 but that the tag is ASN_EOC.

File: wolfcrypt/src/asn.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/asn.c:1495]: Use ASN_INDEF_LENGTH instead of 0x80.

File: wolfcrypt/src/asn.c [hunk 4]
```diff
@@ -2131,20 +2143,31 @@ int GetLength_ex(const byte* input, word32* inOutIdx, int* len, word32 maxIdx,
             minLen = 1 << ((bytes - 1) * 8);
         }
 
-        /* Check the number of bytes required are available. */
-        if ((idx + bytes) > maxIdx) {
-            WOLFSSL_MSG("GetLength - bad long length");
+        /* Check the number of bytes required are available.  At least the
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:2146]: Change check above b == 0x80 to be b == ASN_INDEF_LENGTH.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/asn.c (hunk 2 above)
Comment (tim-weller-wolfssl) [reply wolfcrypt/src/asn.c:1495]: Hi @SparkiDev!  I believe I let the scope of the fix grow beyond just avoiding the out-of-bound access mentioned in the ZenDesk ticket, to start touching on adding support for indefinite-length BER encoding.  I will pull the scope of the changes here back to just avoiding the OoB access (i.e. erroring out instead), and leave the addition of indefinite-length encoding for a later effort (by those who are more in-tune with the ASN.1 logic!).   

I'll leave in as many of your suggested changes as I can without going down the indef-len path.  Thank-you for your time and guidance.  I plan to have this (Draft) PR updated shortly with minimal, OoB-focused changeset.

File: wolfcrypt/src/asn.c (hunk 4 above)
Comment (tim-weller-wolfssl) [reply wolfcrypt/src/asn.c:2146]: Certainly!

File: wolfcrypt/src/asn.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (tim-weller-wolfssl) [reply wolfcrypt/src/asn.c:1028]: Certainly!

File: wolfcrypt/src/asn.c (hunk 3 above)
Comment (tim-weller-wolfssl) [reply wolfcrypt/src/asn.c]: This change has been reverted / removed with the latest PR change-set.

**Review by dgarske — APPROVED** [review]

File: tests/api.c [hunk 5]
```diff
@@ -25651,9 +25651,14 @@ static int test_wc_ecc_rs_to_sig(void)
             ret = WOLFSSL_FATAL_ERROR;
         }
     }
-
     if (ret == 0) {
-        ret = wc_ecc_sig_to_rs(sig, siglen, r, &rlen, s, &slen);
+       ret = wc_ecc_sig_to_rs(sig, siglen, r, &rlen, s, &slen);
```
Comment (SparkiDev) [inline tests/api.c:25655]: This test case is meant to use a good signature.
It looks like the FIPS code is modifying sig in a bad case.
Move this up to after the original wc_ecc_rs_to_sig().

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: tests/api.c (hunk 5 above)
Comment (tim-weller-wolfssl) [reply tests/api.c:25655]: That's a great resolution to the problem!  the change will be in shortly.  Thanks!

**Review by SparkiDev — APPROVED** [review]

**Comment by tim-weller-wolfssl** [comment]
retest this please

**Review by dgarske — APPROVED** [review]

**Pull request merged**